ecl/hql/hqlgram2.cpp

```diff
@@ -12329,30 +12335,30 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
         throw;
     }
     OwnedHqlExpr parsed = scope->lookupSymbol(name, LSFsharedOK|LSFnoreport, ctx);
-    bool canCache = false;
-    bool isMacro = false;
-    bool neverSimplify = false;
-    if (parsed)
-    {
-        canCache = (parsed->getOperator() != no_forwardscope) && !alreadySimplified ;
-        isMacro = parsed->isMacro();
-        neverSimplify = isMacro || ctx.neverSimplify(fullName);
-    }
     ctx.incrementAttribsProcessed();
-    if (canCache && (ctx.syntaxChecking() || ctx.hasCacheLocation()))
+
+    if (parsed && !alreadySimplified)
     {
-        OwnedHqlExpr simplified = neverSimplify ? nullptr : createSimplifiedDefinition(parsed);
-        if (ctx.hasCacheLocation())
-            attrCtx.createCache(simplified, isMacro);
-        if (simplified)
+        const bool canCache = parsed->getOperator() != no_forwardscope;
+        if (canCache && (ctx.hasCacheLocation() || ctx.syntaxChecking()))
         {
-            ctx.incrementAttribsSimplified();
+            const bool isMacro = parsed->isMacro();
+            OwnedHqlExpr simplified;
+            if (!isMacro && !ctx.ignoreSimplified() &&
+                (ctx.hasCacheLocation() || ctx.checkSimpleDef() || ctx.syntaxChecking()) && ctx.neverSimplify(fullName))
```

Comment:
The test ctx.hasCacheLocation is not quite right.

It should probably be

(ctx.hasCacheLocation() && (!cacheIsUpTodate() || ctx.regenerateCache))

The difference is whether a new simplified definition is generated if the cache is already up to date.  I assume the cost of calculating the cache is up to date is cheaper than creating the simplified definition.

Reply:
If the cache is up to date, alreadySimplified will be true.  So, the test in line 12,340 should mean that this section of code will not be executed.